File: scripts/check_spec_anchors.pl
#!/usr/bin/env perl
# Verify every `spec/<path>.md § _Section_` reference in workspace
# source resolves to a real heading in that spec file.
#
# Per-comment-block, position-aware:
#
# - Contiguous `//`, `///`, `//!` lines are folded into one logical
#   block (prefix stripped, joined with spaces) so wrapped anchors
#   like `§ _Update\n//   model_` are seen as one logical token.
# - Each `§ _Section_` is paired with the closest preceding
#   `spec/<path>.md` token in the same block; if none, the most
#   recent mention from the previous 30 source lines carries.
# - Slash continuations `§ _A_ / _B_ / _C_` treat _B_ and _C_ as
#   sibling sections under the same spec file as _A_.
# - Adjacent identical `§ _X_ § _X_` is reported as a duplicate
#   regression (the Round 3 bulk sweep collapsed two distinct
#   references into the same anchor in three places).
# - Heading match is exact — no substring fallback, no near-match.
#
# Run from the repo root: `perl scripts/check_spec_anchors.pl`.
# Exit 0 when every reference resolves; exit 1 with a grouped
# report otherwise.

use strict;
use warnings;
use File::Find;
use File::Spec;

my $root = `git rev-parse --show-toplevel`;
chomp $root;
chdir $root or die "chdir $root: $!";

# Build heading index: { spec_path => { heading => 1 } }
my %headings;
find({
    wanted => sub {
        return unless /\.md\z/ && -f $File::Find::name;
        my $path = File::Spec->abs2rel($File::Find::name, $root);
        open my $fh, '<', $File::Find::name or return;
        while (my $line = <$fh>) {
            if ($line =~ /^\#{1,6}\s+(.+?)\s*$/) {
                $headings{$path}{$1} = 1;
            }
        }
    },
    no_chdir => 1,
}, 'spec');

# Token regex: spec path | primary § _X_ | slash continuation / _Y_.
# The closing underscore must be followed by whitespace, punctuation,
# or end of input so identifiers like `forward_client_ip` aren't eaten.
my $token = qr{
    (spec/[A-Za-z0-9_/-]+\.md)                                   # 1: spec path
    |
    § \s+ _ ([^_\n][^_\n]*?) _ (?=[\s.,;:)\]/`*]|\z)             # 2: primary section
    |
    / \s+ _ ([^_\n][^_\n]*?) _ (?=[\s.,;:)\]/`*]|\z)             # 3: continuation
}x;

my $total = 0;
my %broken;  # "kind\tfile\tsection" => [sites]

my @sources;
find({
    wanted => sub {
        return unless /\.rs\z/ && -f $File::Find::name;
        return if $File::Find::name =~ m{/target/};
        push @sources, File::Spec->abs2rel($File::Find::name, $root);
    },
    no_chdir => 1,
}, 'crates');

for my $src (sort @sources) {
    open my $fh, '<', $src or next;
    my @lines = <$fh>;

    # Group contiguous comment lines into blocks. Each block carries
    # the joined text plus a parallel array mapping the joined-text
    # offset back to the source line number for reporting.
    my @blocks;
    my (@cur_text, @cur_offsets);
    for my $i (0 .. $#lines) {
        if ($lines[$i] =~ m{^\s*//[!/]?[ \t]?(.*?)\s*$}) {
            my $body = $1;
            push @cur_text, $body;
            # Map every character of $body (plus the joining space) to
            # the line index. The trailing space joiner uses the same
            # line so `§\n//   _name_` resolves to the line carrying §.
            push @cur_offsets, ($i) x (length($body) + 1);
        } else {
            if (@cur_text) {
                push @blocks, {
                    text => join(' ', @cur_text),
                    offset_to_line => [@cur_offsets],
                };
                (@cur_text, @cur_offsets) = ();
            }
        }
    }
    if (@cur_text) {
        push @blocks, {
            text => join(' ', @cur_text),
            offset_to_line => [@cur_offsets],
        };
    }

    my ($carry_path, $carry_line) = (undef, -100);
    for my $block (@blocks) {
        my $text = $block->{text};
        my $o2l = $block->{offset_to_line};
        my $first_line = $o2l->[0] // 0;
        my $current = ($first_line - $carry_line <= 30) ? $carry_path : undef;
        my $last_section;

        while ($text =~ /$token/g) {
            my $match_pos = $-[0];
            my $line_num = ($o2l->[$match_pos] // $first_line) + 1;
            if (defined $1) {
                $current = $1;
                $carry_path = $current;
                $carry_line = $o2l->[$match_pos] // $first_line;
                $last_section = undef;
            } elsif (defined $2 || defined $3) {
                my $sec = defined $2 ? $2 : $3;
                $sec =~ s/\s+/ /g;
                $sec =~ s/\.+\z//;
                $total++;
                if (defined $2 && defined $last_section && $last_section eq $sec) {
                    push @{ $broken{"DUP\t" . ($current // '<no-file>') . "\t$sec"} },
                        "$src:$line_num";
                }
                $last_section = $sec if defined $2;
                if (!defined $current) {
                    push @{ $broken{"NOFILE\t<no-file>\t$sec"} }, "$src:$line_num";
                    next;
                }
                if (!exists $headings{$current}) {
                    push @{ $broken{"MISSFILE\t$current\t$sec"} }, "$src:$line_num";
                    next;
                }
                if (!$headings{$current}{$sec}) {
                    push @{ $broken{"MISSHEAD\t$current\t$sec"} }, "$src:$line_num";
                    next;
                }
            }
        }
    }
}

my $broken_count = 0;
$broken_count += scalar @$_ for values %broken;

print "Total scanned: $total\n";
print "Broken: $broken_count\n";

if ($broken_count) {
    print "\n";
    for my $key (sort {
        @{ $broken{$b} } <=> @{ $broken{$a} } || $a cmp $b
    } keys %broken) {
        my ($kind, $sf, $sec) = split /\t/, $key, 3;
        my @sites = @{ $broken{$key} };
        my $plural = @sites == 1 ? '' : 's';
        my $tag = {
            DUP      => 'duplicate',
            NOFILE   => 'no spec file in scope',
            MISSFILE => 'spec file not found',
            MISSHEAD => 'heading not in spec file',
        }->{$kind} // $kind;
        printf "  [%s] %s § _%s_  (%d site%s)\n",
            $tag, $sf, $sec, scalar(@sites), $plural;
        print "    $_\n" for @sites;
    }
    exit 1;
}
exit 0;

File: crates/xtask/src/check_spec_anchors.rs
// Verify spec-anchor references in workspace source resolve to real
// headings in their target spec file. Anchor syntax is documented in
// spec/conventions.md; this checker enforces it.
//
// The matcher is per-comment-block and position-aware:
//
// - Contiguous line-comment lines are folded into one logical block
//   (prefix stripped, joined with spaces) so anchor tokens that wrap
//   across two lines are seen as one logical token.
// - Each section anchor is paired with the closest preceding spec
//   path mention in the same block; if none, the most recent mention
//   from the previous 30 source lines carries forward.
// - Slash continuations after a primary anchor are treated as sibling
//   sections under the same spec file as the primary.
// - Two same-section primary anchors with only whitespace between
//   them are reported as a duplicate regression. Same-section
//   references that are separated by prose (or by a different anchor)
//   are not flagged — they're assumed intentional cross-references.
// - Heading match is exact — no substring fallback, no near-match.

use std::collections::{BTreeMap, HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use fancy_regex::Regex;

use crate::workspace;

pub(crate) fn run() -> Result<()> {
	let root = workspace::root()?;

	let headings = build_heading_index(&root.join("spec"), &root)?;
	let sources = collect_rust_sources(&root.join("crates"), &root)?;

	let heading_re = Regex::new(r"^\s*//[!/]?[ \t]?(.*?)\s*$").unwrap();
	let token_re = Regex::new(concat!(
		r"(spec/[A-Za-z0-9_/-]+\.md)",
		r"|",
		r"§\s+_([^_\n][^_\n]*?)_(?=[\s.,;:)\]/`*]|\z)",
		r"|",
		r"/\s+_([^_\n][^_\n]*?)_(?=[\s.,;:)\]/`*]|\z)",
	))
	.unwrap();

	let mut total: usize = 0;
	let mut broken: BTreeMap<BrokenKey, Vec<String>> = BTreeMap::new();

	for src in &sources {
		let abs = root.join(src);
		let content = fs::read_to_string(&abs).with_context(|| format!("reading {}", abs.display()))?;
		scan_file(src, &content, &heading_re, &token_re, &headings, &mut total, &mut broken)?;
	}

	let broken_count: usize = broken.values().map(Vec::len).sum();
	println!("Total scanned: {total}");
	println!("Broken: {broken_count}");

	if broken_count == 0 {
		return Ok(());
	}

	println!();
	let mut entries: Vec<_> = broken.iter().collect();
	entries.sort_by(|(ka, va), (kb, vb)| vb.len().cmp(&va.len()).then_with(|| ka.cmp(kb)));
	for (key, sites) in entries {
		let plural = if sites.len() == 1 { "" } else { "s" };
		println!(
			"  [{}] {} § _{}_  ({} site{plural})",
			key.kind.label(),
			key.spec_file,
			key.section,
			sites.len(),
		);
		for site in sites {
			println!("    {site}");
		}
	}
	bail!("{broken_count} broken spec anchor reference(s)");
}

#[derive(Eq, PartialEq, Ord, PartialOrd)]
struct BrokenKey {
	kind: BrokenKind,
	spec_file: String,
	section: String,
}

#[derive(Eq, PartialEq, Ord, PartialOrd)]
enum BrokenKind {
	Duplicate,
	NoFile,
	MissingFile,
	MissingHeading,
}

impl BrokenKind {
	fn label(&self) -> &'static str {
		match self {
			Self::Duplicate => "duplicate",
			Self::NoFile => "no spec file in scope",
			Self::MissingFile => "spec file not found",
			Self::MissingHeading => "heading not in spec file",
		}
	}
}

fn build_heading_index(spec_dir: &Path, root: &Path) -> Result<HashMap<String, HashSet<String>>> {
	let head_re = Regex::new(r"^#{1,6}\s+(.+?)\s*$").unwrap();
	let mut headings: HashMap<String, HashSet<String>> = HashMap::new();
	for path in walk_files(spec_dir, "md")? {
		let rel = path.strip_prefix(root).unwrap_or(&path).to_string_lossy().into_owned();
		let content =
			fs::read_to_string(&path).with_context(|| format!("reading {}", path.display()))?;
		let entry = headings.entry(rel).or_default();
		for line in content.lines() {
			if let Ok(Some(c)) = head_re.captures(line)
				&& let Some(h) = c.get(1)
			{
				entry.insert(h.as_str().to_string());
			}
		}
	}
	Ok(headings)
}

fn collect_rust_sources(crates_dir: &Path, root: &Path) -> Result<Vec<PathBuf>> {
	let mut out = Vec::new();
	for path in walk_files(crates_dir, "rs")? {
		if path.components().any(|c| c.as_os_str() == "target") {
			continue;
		}
		let rel = path.strip_prefix(root).unwrap_or(&path).to_path_buf();
		out.push(rel);
	}
	out.sort();
	Ok(out)
}

fn walk_files(dir: &Path, ext: &str) -> Result<Vec<PathBuf>> {
	let mut out = Vec::new();
	walk_into(dir, ext, &mut out)?;
	Ok(out)
}

fn walk_into(dir: &Path, ext: &str, out: &mut Vec<PathBuf>) -> Result<()> {
	if !dir.is_dir() {
		return Ok(());
	}
	for entry in fs::read_dir(dir).with_context(|| format!("reading dir {}", dir.display()))? {
		let entry = entry?;
		let path = entry.path();
		if path.is_dir() {
			walk_into(&path, ext, out)?;
		} else if path.extension().and_then(|s| s.to_str()) == Some(ext) {
			out.push(path);
		}
	}
	Ok(())
}

struct CommentBlock {
	text: String,
	// One offset per byte of `text` (plus one trailing slot for the
	// joiner), each entry the 0-indexed source line that byte came
	// from. `text.len() <= offset_to_line.len()`.
	offset_to_line: Vec<usize>,
}

fn group_comment_blocks(content: &str, comment_re: &Regex) -> Vec<CommentBlock> {
	let mut blocks: Vec<CommentBlock> = Vec::new();
	let mut text = String::new();
	let mut offsets: Vec<usize> = Vec::new();
	let mut started = false;

	for (i, line) in content.lines().enumerate() {
		let captured = comment_re
			.captures(line)
			.ok()
			.flatten()
			.and_then(|c| c.get(1))
			.map(|m| m.as_str().to_string());
		if let Some(body) = captured {
			if started {
				text.push(' ');
			}
			text.push_str(&body);
			for _ in 0..body.len() {
				offsets.push(i);
			}
			offsets.push(i);
			started = true;
		} else if started {
			blocks.push(CommentBlock {
				text: std::mem::take(&mut text),
				offset_to_line: std::mem::take(&mut offsets),
			});
			started = false;
		}
	}
	if started {
		blocks.push(CommentBlock { text, offset_to_line: offsets });
	}
	blocks
}

fn scan_file(
	src: &Path,
	content: &str,
	comment_re: &Regex,
	token_re: &Regex,
	headings: &HashMap<String, HashSet<String>>,
	total: &mut usize,
	broken: &mut BTreeMap<BrokenKey, Vec<String>>,
) -> Result<()> {
	let blocks = group_comment_blocks(content, comment_re);
	// Carry the most recent spec path mention forward across blocks
	// when the gap is at most 30 source lines; `None` means "no
	// reference to carry yet."
	let mut carry: Option<(String, usize)> = None;
	let src_str = src.to_string_lossy();

	for block in &blocks {
		let first_line = block.offset_to_line.first().copied().unwrap_or(0);
		let mut current = match &carry {
			Some((path, line)) if first_line.saturating_sub(*line) <= 30 => Some(path.clone()),
			_ => None,
		};
		// Tracks the most recent primary anchor: section name plus the
		// byte offset right after its closing `_`. Used to flag adjacent
		// duplicates only when nothing but whitespace sits between the
		// two matches.
		let mut last_primary: Option<(String, usize)> = None;

		for cap in token_re.captures_iter(&block.text) {
			let cap = cap.context("regex capture iteration failed")?;
			let whole = cap.get(0).unwrap();
			let pos = whole.start();
			let line_num = block.offset_to_line.get(pos).copied().unwrap_or(first_line) + 1;

			if let Some(spec) = cap.get(1) {
				let path = spec.as_str().to_string();
				current = Some(path.clone());
				let at_line = block.offset_to_line.get(pos).copied().unwrap_or(first_line);
				carry = Some((path, at_line));
				last_primary = None;
				continue;
			}

			let primary = cap.get(2);
			let cont = cap.get(3);
			let Some(matched) = primary.or(cont) else {
				continue;
			};

			let mut sec = matched.as_str().replace(['\n', '\t'], " ");
			while sec.contains("  ") {
				sec = sec.replace("  ", " ");
			}
			sec = sec.trim_end_matches('.').to_string();

			*total += 1;

			if primary.is_some() {
				if let Some((prev_sec, prev_end)) = &last_primary
					&& prev_sec == &sec
					&& block.text[*prev_end..pos].chars().all(char::is_whitespace)
				{
					broken
						.entry(BrokenKey {
							kind: BrokenKind::Duplicate,
							spec_file: current.clone().unwrap_or_else(|| "<no-file>".into()),
							section: sec.clone(),
						})
						.or_default()
						.push(format!("{src_str}:{line_num}"));
				}
				last_primary = Some((sec.clone(), whole.end()));
			}

			let Some(spec_path) = current.as_ref() else {
				broken
					.entry(BrokenKey {
						kind: BrokenKind::NoFile,
						spec_file: "<no-file>".into(),
						section: sec.clone(),
					})
					.or_default()
					.push(format!("{src_str}:{line_num}"));
				continue;
			};

			let Some(spec_headings) = headings.get(spec_path) else {
				broken
					.entry(BrokenKey {
						kind: BrokenKind::MissingFile,
						spec_file: spec_path.clone(),
						section: sec.clone(),
					})
					.or_default()
					.push(format!("{src_str}:{line_num}"));
				continue;
			};

			if !spec_headings.contains(&sec) {
				broken
					.entry(BrokenKey {
						kind: BrokenKind::MissingHeading,
						spec_file: spec_path.clone(),
						section: sec.clone(),
					})
					.or_default()
					.push(format!("{src_str}:{line_num}"));
			}
		}
	}
	Ok(())
}
